src/findlib.ml

```diff
@@ -180,107 +151,93 @@ module Package_not_available = struct
-        type nonrec t = t
-        let key t = t.package
-        let deps t () =
-          match t.reason with
-          | Not_found | Hidden -> []
-          | Dependencies_unavailable l -> l
-      end)
-
-  let all_names ts =
-    let rec loop acc ts =
-      List.fold_left ts ~init:acc ~f:(fun acc t ->
-        if String_set.mem t.package acc then
-          acc
-        else
-          let acc = String_set.add t.package acc in
-          match t.reason with
-          | Not_found | Hidden -> acc
-          | Dependencies_unavailable ts -> loop acc ts)
-    in
-    loop String_set.empty ts |> String_set.elements
-
-  let top_closure ts =
-    match Closure.top_closure () ts with
-    | Ok ts   -> ts
-    | Error _ ->
-      code_errorf "Findlib.Package_not_available.top_sort got a cycle:\n%s"
-        (all_names ts
-         |> List.map ~f:(sprintf "- %s")
-         |> String.concat ~sep:"\n")
 
   let explain ppf reason =
     match reason with
     | Not_found ->
       Format.fprintf ppf "not found"
     | Hidden ->
       Format.fprintf ppf "hidden (unsatisfied 'exist_if')"
-    | Dependencies_unavailable deps ->
-      Format.fprintf ppf
-        "@[<2>unavailable dependencies:@ %t@]"
-        (fun ppf ->
-           match deps with
-           | [] -> ()
-           | t :: rest ->
-             Format.fprintf ppf "%s" t.package;
-             List.iter rest ~f:(fun t ->
-               Format.fprintf ppf ",@ %s" t.package))
 end
 
-type present_or_not_available =
-  | Present       of Package.t
-  | Not_available of Package_not_available.t
+module External_dep_conflicts_with_local_lib = struct
+  type t =
+    { package             : string
+    ; required_by         : With_required_by.Entry.t
+    ; required_locally_in : With_required_by.Entry.t list
+    ; defined_locally_in  : Path.t
+    }
+end
+
+module Dependency_cycle = struct
+  type t =
+    { cycle       : string list
+    ; required_by : With_required_by.Entry.t list
+    }
+end
+
+type error =
+  | Package_not_available
+    of Package_not_available.t
+  | External_dep_conflicts_with_local_lib
+    of External_dep_conflicts_with_local_lib.t
+  | Dependency_cycle
+    of Dependency_cycle.t
 
 type t =
   { stdlib_dir    : Path.t
   ; path          : Path.t list
   ; builtins      : Meta.t String_map.t
   ; packages      : (string, present_or_not_available) Hashtbl.t
+  (* Cache the result of [closure]. A key is the list of package
+     unique identifiers. *)
+  ; closure_cache : (int list, (package list, error) result) Hashtbl.t
+  }
+
+and package =
+  { name                     : string
+  ; unique_id                : int
+  ; dir                      : Path.t
+  ; version                  : string
+  ; description              : string
+  ; archives                 : Path.t list Mode.Dict.t
+  ; plugins                  : Path.t list Mode.Dict.t
+  ; jsoo_runtime             : string list
+  ; mutable requires         : deps
+  ; mutable ppx_runtime_deps : deps
```

Comment:
Is it possible to just make these 2 fields lazy? It would prevent them from being assigned more than once.

Reply:
Indeed, I changed the code